packages/flutter/lib/src/cupertino/segmented_control.dart

```diff
@@ -0,0 +1,513 @@
+  /// appear as selected.
+  SegmentedControl({
+    Key key,
+    @required this.children,
+    @required this.onValueChanged,
+    this.groupValue,
+  })  : assert(children != null),
+        assert(children.length >= 2),
+        assert(onValueChanged != null),
+        assert(groupValue == null || children.keys.any((T child) => child == groupValue)),
+        super(key: key);
+
+  /// The identifying keys and corresponding widget values in the
+  /// segmented control.
+  ///
+  /// This attribute must not be null. The length of the list must be greater
+  /// than or equal to two. This attribute must be an ordered [Map] such as a
+  /// [LinkedHashMap].
+  final Map<T, Widget> children;
+
+  /// The identifier of the widget that is currently selected.
+  ///
+  /// This must be one of the keys in the [Map] of [children].
+  /// If this attribute is null, no widget will be initially selected.
+  final T groupValue;
+
+  /// The callback that is called when a new option is tapped.
+  ///
+  /// This attribute must not be null.
+  ///
+  /// The segmented control passes the newly selected widget's associated key
+  /// to the callback but does not actually change state until the parent
+  /// widget rebuilds the segmented control with the new [groupValue].
+  ///
+  /// The callback provided to [onValueChanged] should update the state of
+  /// the parent [StatefulWidget] using the [State.setState] method, so that
+  /// the parent gets rebuilt; for example:
+  ///
+  /// ## Sample code
+  ///
+  /// ```dart
+  /// class SegmentedControlExample extends StatefulWidget {
+  ///   @override
+  ///   State createState() => new SegmentedControlExampleState();
+  /// }
+  ///
+  /// class SegmentedControlExampleState extends State<SegmentedControlExample> {
+  ///   final Map<int, Widget> children = const {
+  ///     0: const Text('Child 1'),
+  ///     1: const Text('Child 2'),
+  ///   };
+  ///
+  ///   int currentValue;
+  ///
+  ///   @override
+  ///   Widget build(BuildContext context) {
+  ///     return new Container(
+  ///       child: new SegmentedControl<int>(
+  ///         children: children,
+  ///         onValueChanged: (int newValue) {
+  ///           setState(() {
+  ///             currentValue = newValue;
+  ///           });
+  ///         },
+  ///         groupValue: currentValue,
+  ///       ),
+  ///     );
+  ///   }
+  /// }
+  /// ```
+  final ValueChanged<T> onValueChanged;
+
+  @override
+  _SegmentedControlState<T> createState() => _SegmentedControlState<T>();
+}
+
+class _SegmentedControlState<T> extends State<SegmentedControl<T>> {
+  int _selectedIndex;
+  int _pressedIndex;
+  Map<T, int> keysToIndices = <T, int>{};
+  @override
+  void initState() {
+    super.initState();
+    for (int index = 0; index < widget.children.keys.length; index += 1) {
+      keysToIndices[widget.children.keys.elementAt(index)] = index;
+    }
+    if (widget.children.keys.contains(widget.groupValue)) {
```

Comment:
Seems like you should be able to just assert this, and instead of doing a search in the keys list, just check and see if widget.groupValue is null or not.  The contract in the constructor already asserts that one of these things will be true, so you shouldn't need to check it again here.